[NUI] Add Widget documents (#988)
Samsung/tizen-docs: docs/application/dotnet/guides/nui/widget.md

> # NUI Widget Application

Author: Excuse me, can you tell me what means "toc file"?

> # NUI Widget Application

Reviewer: @tscholb, since it is a new file, we require to update the table of contents (toc) file.

https://github.com/Samsung/tizen-docs/blob/d2a99fcbf74263f2f056000359a03db1f9413dcb/docs/application/toc_all.md

Hope this helps.

> # NUI Widget Application

Author: Thank you for your kind explanation.
I will update it  around next week.

You can use [NUIWidgetApplication](https://samsung.github.io/TizenFX/latest/api/Tizen.NUI.NUIWidgetApplication.html) to create Tizen.NET widget applications, which are small versions of Tizen.NET applications on the home screen. 

> You can launch these applications on the home screen when you select them in the Add Widget menu.

Reviewer: Please use bold tag for menu name

```suggestion
You can use [NUIWidgetApplication](https://samsung.github.io/TizenFX/latest/api/Tizen.NUI.NUIWidgetApplication.html) to create Tizen.NET widget applications, which are small versions of Tizen.NET applications on the home screen. You can launch these applications on the home screen when you select them in the **Add Widget** menu.
```

![NUI widget application](./media/dali_widget_application.png)

The main features of NUIWidgetApplication include:

- Create widget applications

  You can [create a NUI widget application](#create-a-nui-widget-application), which usually is composed of one process having the main loop. The NUIWidgetApplication framework enables you to create a widget instance that shares the same resources within the process. The widget application can also share data with other applications.

- Manage the life-cycle

  You can manage the life-cycle of widget instances through callback methods that are triggered as the instance states are changed.

- Manage widget instances
  
  Whenever a widget instance is requested, the framework creates one.
  You can manage the widget instances by updating or destroying them, or by retrieving information about them.

- Create the widget UI

  The widget application can create a user interface that appears on the home screen or viewer.

    > **Note**
    >
    > The UI of widget application limits scroll actions to provide a better user experience. It is recommended that you design a UI layout within the given screen size.
    >

>     > To draw the layout, you must use the [`Window`](https://developer.tizen.org/dev-guide/csapi/api/Tizen.NUI.Window.html) instance received from the `OnCreate` callback. If you create additional windows, the stack of widget application windows gets corrupted. This is because the platform handles the widget application window in a particular way.

Reviewer: Global: As per the new guidelines for style, please use the updated Note tag.
Reference: https://github.com/Samsung/tizen-docs/blob/master/styleguide/custom-style.md

> [!NOTE]
> The UI of widget application limits scroll actions to provide a better user experience. It is recommended that you design a UI layout within the given screen size.
>
> To draw the layout, you must use the [`Window`](https://developer.tizen.org/dev-guide/csapi/api/Tizen.NUI.Window.html) instance received from the `OnCreate` callback. If you create additional windows, the stack of widget application windows gets corrupted. This is because the platform handles the widget application window in a particular way.

## Application and Viewer

[`NUIWidgetApplication`](https://samsung.github.io/TizenFX/latest/api/Tizen.NUI.NUIWidgetApplication.html) makes diverse class instances whenever [`WidgetView`](https://samsung.github.io/TizenFX/latest/api/Tizen.NUI.WidgetView.html) requests for a widget instance.

The widget instance has its own life-cycle similar to that of the widget application. However, the widget instance is an object created by the widget viewer application. Many widget instances can be running on a widget application process.

>**Note**
>

> >The case to use many widget instances in one widget application is known as multi-instance. In some devices, the multi-instance may not be supported. If a device does not support multi-instance, an error message is displayed.

Reviewer: Global: As per the new guidelines for style, please use the updated Note tag.
Reference: https://github.com/Samsung/tizen-docs/blob/master/styleguide/custom-style.md

>[!Note]
>The case to use many widget instances in one widget application is known as multi-instance. In some devices, the multi-instance may not be supported. If a device does not support multi-instance, an error message is displayed.

> # NUI Widget Application

Reviewer: @tscholb, please incorporate this comment.

> [NUIWidgetApplication](https://samsung.github.io/TizenFX/latest/api/Tizen.NUI.NUIWidgetApplication.html)  is to show a small version of application on home screen.

Reviewer 2: > Suggestion: Please provide some more information in the introduction such as why do we use this, what is the purpose, when do we use, and so on.
> 
> For example,
> You can use [NUIWidgetApplication](https://samsung.github.io/TizenFX/latest/api/Tizen.NUI.NUIWidgetApplication.html) to create Tizen.NET widget applications, which are small versions of Tizen .NET applications on the home screen. You can launch these applications on the home screen when you select them in the **Add Widget** menu.
> 
> Please verify. This is just an example. Kindly elaborate the introduction.

"You can launch these applications on the home screen when you select them in the Add Widget menu." is not suitable for this guide because your rephase assumes "Add Widget" menu on the home screen.

I think the feature or definition of this module is better than "you can...", as the start wording.

> [NUIWidgetApplication](https://samsung.github.io/TizenFX/latest/api/Tizen.NUI.NUIWidgetApplication.html)  is to show a small version of application on home screen.

Reviewer: @journey2w, thank you for the correction.
Could we start with:

NUIWidgetApplication helps you to create Tizen.NET widget applications, which are .....

Secondly, request the author to please provide an introduction.

**Figure: NUI widget application**

> ![NUI widget application](./media/dali_widget_application.png)

Reviewer 2: > The image is not appearing. Also, was unable to find the image in the path:
> https://github.com/Samsung/tizen-docs/tree/master/docs/application/dotnet/guides/nui/media
> 
> Please verify the link

Don't leave duplecated comments.

<br>
The main widget application features include:

- Creating widget applications

>   You can [Create a NUI Widget Application](#create-a-nui-widget-application)

Reviewer 2: > Suggestion:
> 
> Can we change to lower case if it is not a Tizen specific naming convention: [create a NUI widget application]?
> 
> Suggest rephrasing to:
> 
> You can [create a NUI widget application](#create-a-nui-widget-application), which usually has one process for maintaining the main loop. Within this process, the framework can a create widget instance that can share the same resources.
> 
> @mijin85cho, incase NUI Widget Application (all upper case) is a Tizen specific naming convention, then suggest adding it to https://github.com/Samsung/tizen-docs/blob/master/styleguide/naming-rules.md

Widget is a normal noun and application also is. I think NUI widget application is correct.

>   You can [Create a NUI Widget Application](#create-a-nui-widget-application)

Reviewer: @journey2w,

Thank you for the clarification.

, which usually has 1 process for maintaining the main loop. Within the process, the framework can be create widget instance that can share the same resources.
  The Widget application can also share data with other applications.

- Managing widget instances
  

>   Whenever a widget instance is requested, the framework creates one.

Reviewer: @tscholb , please incorporate @journey2w  suggetsion.

Reference:https://github.com/Samsung/tizen-docs/pull/988/files#r363568303

  You can manage the widget instances by updating or destroying them, or by retrieving information about them.

> - Managing the life-cycle

Reviewer: Global, please avoid using Gerund form.

Please change to:

- Manage life-cycle

Reviewer 2: > Global, please avoid using Gerund form.
> 
> Please change to:
> 
> * Manage life-cycle

We are using a verb as the section title. However, This is a list. I think managing the life-cycle is correct because the gerund is used as the list items.

We've never guided you to avoid using a gerund for list items, have we?

Reviewer: @journey2w, thank you for pointing this out.

Request the author to please ignore this comment wherever applicable and maintain the original content.

>   You can manage the widget instance life-cycle through callback methods that are triggered as the instance state changes.

Reviewer: Suggetsion: Will it be good to provide a link to the widget life-cycle ?https://docs.tizen.org/application/dotnet/guides/applications/widget-app#instance

Please ignore this comment if invalid.

> - Creating the widget UI

Reviewer: Global, please avoid using Gerund form.

Please change to:

- Creating widget UI

Reviewer 2: > Global, please avoid using Gerund form.
> 
> Please change to:
> 
> * Create widget UI

Same as above.

>   The widget application can draw a UI on the Home screen or viewer

Reviewer: Suggest changing to:

The widget application can draw a UI on the home screen or viewer.

    > **Note**
    >
    > The widget application UI has a limitation with the scroll action to provide a better user experience. Design the widget UI to display all the information within the given area of the screen points.
    >

>     > To draw the UI, you must use the [Window](https://developer.tizen.org/dev-guide/csapi/api/Tizen.NUI.Window.html) received from the OnCreate Callback.  Do not create additional windows. A stack of widget application windows gets corrupted, because the platform handles the widget application window in a special way.

Reviewer: Suggest rephrasing to:

> To draw the layout, you must use the [Window](https://developer.tizen.org/dev-guide/csapi/api/Tizen.NUI.Window.html) received from the `OnCreate` callback.  If you create additional windows, the stack of widget application windows gets corrupted. This is because the platform handles the widget application window in a particular way.

## Application and Viewer

[NUIWidgetApplication](https://samsung.github.io/TizenFX/latest/api/Tizen.NUI.NUIWidgetApplication.html) can make diverse class instances whenever [WidgetView](https://samsung.github.io/TizenFX/latest/api/Tizen.NUI.WidgetView.html) request for a widget instance.<br>

> The widget instance has its own life-cycle similar to the widget application. However, the widget instance is only an object shown by the widget viewer applications. Many widget instances can be running on the same widget application process.

Reviewer: @tscholb, please incorporate https://github.com/Samsung/tizen-docs/pull/988/files#r363572973 comment.

> The widget instance has its own life-cycle similar to the widget application. However, the widget instance is only an object shown by the widget viewer applications. Many widget instances can be running on the same widget application process.

Reviewer: @tscholb, please incorporate https://github.com/Samsung/tizen-docs/pull/988/files#r363572973 comment.

**Figure: NUI widget flow**

![NUI widget flow](./media/WidgetInstanceFlow.png)

`Widget` class provides interface for creating custom widget.

The table list the main functions to manage widget instance:

**Table: Main functions**

| Function           | Description                                                                                   |
|--------------------|-----------------------------------------------------------------------------------------------|
| `OnCreate()`       | This function is called after the widget instance is created.                                 |
| `OnTerminate()`    | This function is called after the widget instance is terminated.                              |
| `OnPause()`        | This function is called when the widget is invisible.                                         |
| `OnResume()`       | This function is called when the widget is visible.                                           |
| `OnResize()`       | This function is called before the widget size is changed.                                    |
| `OnUpdate()`       | This function is called when an event for updating the widget is received.                    |

> | `SetContentInfo()` | This function sends data about current state of widget instance to a widget viewer application. |

Reviewer 2: > Please add articles. Suggest rephrasing to:
> 
> This function sends data about the current state of the widget instance to a widget viewer application.
> 
> Reference: https://github.com/Samsung/tizen-docs/blob/master/styleguide/naming-rules.md

The current phrase is no problem.

> | `SetContentInfo()` | This function sends data about current state of widget instance to a widget viewer application. |

Reviewer: @journey2w, I agree that the original sentence was correct. 
I suggested adding articles to maintain consistency with the other descriptions in this table.
Let me know how if we should consider this comment.

            <support-size preview="Widget.png">4x4</support-size>
          < /widget-class>
          ```
      - Using constructor as follows
        ```csharp
        static void Main(string[] args)
        {
           Dictionary<System.Type, string> widgetSet = new Dictionary<Type, string>();
            widgetSet.Add(typeof(MyWidget), "second@org.tizen.example.WidgetTemplate");
            var app = new Program(widgetSet);
            app.Run(args);
        }    
        ```
      
    

4. Define your widget class, which is inherited from the [Widget](https://samsung.github.io/TizenFX/latest/api/Tizen.NUI.Widget.html)

   ```csharp
   class MyWidget : Widget
   ```

   

5. Overide the event callback methods of your new class:

  - The `OnCreate()` callback is triggered when the widget instance is created.

     Initialize resources for this widget instance and draw the UI.

   
    ```csharp
    class MyWidget : Tizen.NUI.Widget
    {
      protected override void OnCreate(string contentInfo, Window window)
      {
        /// Create the UI
        /// ....
        base.OnCreate(contentInfo, window);
      }
    }
    ```

> 

Reviewer: @tscholb,

Please  incorporate comments given by @journey2w https://github.com/Samsung/tizen-docs/pull/988/files#r363589083

      ```csharp
      protected override void OnTerminate(string contentInfo, TerminationType type)
      ```

  - The `OnPause()` callback is triggered when the widget instance is paused.

      Take the necessary actions when the widget instance becomes invisible. The framework can destroy a paused widget instance.
      ```csharp
      protected override void OnPause()
      ```

  - The `OnResume()` callback is triggered when the widget instance is resumed.

      Take the necessary actions when the widget instance becomes visible.

      ```csharp
      protected override void OnResume()
      ```

  - The `OnResize()` callback is triggered when the widget instance is resized.

      Take the necessary actions to accommodate the new size.
      ```csharp
      protected override void OnResize(Window window)
      ```

  - The `OnUpdate()` callback is triggered when a widget update event is received. 

      Take the necessary actions for the widget update. If the isForce parameter is true, the widget can be updated even in the pause state.
      ```csharp
      protected override void OnUpdate(string contentInfo, int force)

>       ```

Reviewer: @tscholb, Please incorporate this suggestion and use code tag for `isForce`

>       ```

Reviewer: @tscholb, please incorporate the comments given earlier: https://github.com/Samsung/tizen-docs/pull/988/files#r363590662

    The widget UI is drawn in the `OnCreate()` callback of your widget class:

    ```csharp
    protected override void OnCreate(string contentInfo, Window window)
    {
        View rootView = new View();
        rootView.BackgroundColor = Color.White;
        rootView.Size2D = window.Size;
        rootView.PivotPoint = PivotPoint.Center;
        window.GetDefaultLayer().Add(rootView);

        TextLabel sampleLabel = new TextLabel("Hello World!");
        sampleLabel.FontFamily = "SamsungOneUI 500";
        sampleLabel.PointSize = 71;
        sampleLabel.TextColor = Color.Black;
        sampleLabel.SizeWidth = 200;
        sampleLabel.PivotPoint = PivotPoint.Center;

        rootView.Add(sampleLabel);
    }
    ```

## Create a NUI Widget View

If necessary, you can create a Widget Viewer that shows UI drawn by Widget

The steps to create a nui widget view:<br>
1. Add privileges for viewer to tizen-manifest.xml

```XML
  <privileges>
    <privilege>http://tizen.org/privilege/widget.viewer</privilege>
    <privilege>http://tizen.org/privilege/appmanager.launch</privilege>
  </privileges>
```
2. Create widget manager and request add widget to widget manager

> - Using the `NUIWidgetApplication(Type type)` constructor, add a widget using widget app ID.

Reviewer 2: Insert 4 spaces in front of this paragraph.

```suggestion
    You can create a widget manager using two different constructors:
    
    - The `NUIWidgetApplication(Type type)` constructor for adding a widget using the widget app ID.
```

```csharp
protected override void OnCreate()
{
  WidgetViewManager widgetViewManager = new WidgetViewManager(this, this.ApplicationInfo.ApplicationId);

  WidgetView myWidgetView = widgetViewManager.AddWidget("org.tizen.example.Widget", encodedBundle, width, height, 0);
  Window.Instance.GetDefaultLayer().Add(myWidgetView);
```

- Using the `NUIWidgetApplication(Dictionary< Type, string > typeInfo) ` constructor, add a widget using class and app ID
```csharp
protected override void OnCreate()
{
  WidgetViewManager widgetViewManager = new WidgetViewManager(this, this.ApplicationInfo.ApplicationId);

  WidgetView myWidgetVie1 = widgetViewManager.AddWidget("second@org.tizen.example.Widget", encodedBundle, width, height, 0);
  Window.Instance.GetDefaultLayer().Add(myWidgetView);

> ```

Reviewer 2: Insert 4 spaces in front of this paragraph.

```suggestion
2. Create a widget manager and request for adding a widget to the widget manager.

    You can create a widget manager using two different constructors:
    
    - The `NUIWidgetApplication(Type type)` constructor for adding a widget using the widget app ID.
       ```csharp
        protected override void OnCreate()
        {
           WidgetViewManager widgetViewManager = new WidgetViewManager(this, this.ApplicationInfo.ApplicationId);
          WidgetView myWidgetView = widgetViewManager.AddWidget("org.tizen.example.Widget", encodedBundle, width, height, 0);
          Window.Instance.GetDefaultLayer().Add(myWidgetView);
          ...  
          }
        ```
    - The `NUIWidgetApplication(Dictionary< Type, string > typeInfo) ` constructor for adding a widget using class and app ID.

        ```csharp
        protected override void OnCreate()
        {
          WidgetViewManager widgetViewManager = new WidgetViewManager(this, this.ApplicationInfo.ApplicationId);
          WidgetView myWidgetVie1 = widgetViewManager.AddWidget("second@org.tizen.example.Widget",         encodedBundle, width, height, 0);
          Window.Instance.GetDefaultLayer().Add(myWidgetView);
          ...
        ```
```
--- 

By the way, the code snippet is the same except values, myWidgetView and myWidgetVie1.

As I understand, the following paragraph is better:
```suggestion

2. Create a widget manager and request for adding a widget to the widget manager.

    You can create a widget manager in `onCreate()` method:
     ```csharp
     protected override void OnCreate()
     {
           WidgetViewManager widgetViewManager = new WidgetViewManager(this, this.ApplicationInfo.ApplicationId);
          WidgetView myWidgetView = widgetViewManager.AddWidget("org.tizen.example.Widget", encodedBundle, width, height, 0);
          Window.Instance.GetDefaultLayer().Add(myWidgetView);
          ...  
      }
